feat: allow user to provide TS program instance in parser options

uniqueiniquity opened the pull request:
`typescript-estree` has complex handling for constructing and managing TS program instances in order to provide type info for semantic rules. However, for one-off invocations on large codebases (e.g., in CI), this functionality isn't valuable, and the overhead is substantial.

Here, we introduce a new parser option `program` that allows users to programmatically provide their own TS program instance that will be used in rules for type info. This is based on [TSLint's functionality](https://github.com/palantir/tslint/blob/285fc1db18d1fd24680d6a2282c6445abf1566ee/src/linter.ts#L121).

In this initial implementation, we only allow for a single program to be provided. However, this could be augmented in the future to support multiple programs.

Fixes https://github.com/typescript-eslint/typescript-eslint/issues/1442

bradzacher commented on packages/typescript-estree/package.json:
```diff
@@ -64,6 +64,9 @@
     "tmp": "^0.2.1",
+    "typescript": "^4.3.2"
+  },
+  "peerDependencies": {
```
We have an implicit peer dependency via the `peerDependenciesMeta` property.

The reason we don't explicitly declare the peer dependency is because on older versions of yarn and npm, they do not understand `peerDependenciesMeta`.
So if we add an explicit peer dep - then they will log warnings at install time.

In the past this annoyed many users 😅 that had our packages installed (but not used) transitively.

TL;DR - remove the explicit peer dep

bradzacher commented on packages/types/package.json:
```diff
@@ -48,5 +48,8 @@
     }
+  },
+  "devDependencies": {
+    "typescript": ">=2.6.0"
```
We use the root package.json to control the TS version installed in the repo.

You can use the `peerDependenciesMeta` to add an optional peer dep on typescript, which we'll need to do to ensure package managers can understand the new requirement

```suggestion
    "typescript": "*"
```

bradzacher commented on packages/types/src/parser-options.ts:
```diff
@@ -1,4 +1,5 @@
 import { Lib } from './lib';
+import { Program } from 'typescript';
```
lets ensure that this isn't ever pulled in as a runtime dep on this package.

```suggestion
import type { Program } from 'typescript';
```

bradzacher commented on packages/typescript-estree/src/create-program/useProvidedProgram.ts:
```diff
@@ -0,0 +1,28 @@
+  if (!astAndProgram) {
+    const errorLines = [
+      '"parserOptions.program" has been provided for @typescript-eslint/parser.',
+      `The file was not found in the provided program instance: ${extra.filePath}`,
```
The filepath should be made relative.
This ensures that error messages are kept is stable and portable.

https://github.com/typescript-eslint/typescript-eslint/blob/fc5f171b1ade2b45a1a4268b6d22926d420282a6/packages/typescript-estree/src/create-program/createProjectProgram.ts#L52-L55

bradzacher commented on packages/typescript-estree/src/parser.ts:
```diff
@@ -264,22 +269,32 @@ function applyParserOptionsToExtra(options: TSESTreeOptions): void {
+    );
+  }
+
+  if (extra.program === null) {
```
nit - just for safety so we can handle the undefined case as well, in case something somewhere is wacky.

```suggestion
  if (extra.program == null) {
```

uniqueiniquity replied:
I made this `if (!extra.program)`

bradzacher commented on packages/typescript-estree/src/parser.ts:
```diff
@@ -264,22 +269,32 @@ function applyParserOptionsToExtra(options: TSESTreeOptions): void {
-    options.project,
-    projectFolderIgnoreList,
-  );
+  if (typeof options.program === 'object') {
```
nit - we don't want to log if the user explicitly passes `null` here.

```suggestion
  if (typeof options.program === 'object' && options.program != null) {
```

uniqueiniquity replied:
I made this `if (options.program && typeof options.program === 'object')`

bradzacher reviewed (changes requested):
i didn't review the tests - just glanced through the code.
looks good! a few comments.

have you tested this in a real-world lint setup?
I'd be interested to know how much this improves the perf for them.
Also would like to triple check that ESLint doesn't mangle the program instance or anything.

uniqueiniquity commented:
@bradzacher I'm about to try it on a large project - will report back alongside addressing your suggestions.

uniqueiniquity commented:
Closing in favor of #3484 - I screwed up my branch casing and couldn't fix it 😫
I've addressed all comments currently on this PR and added perf results on the new PR - spoiler alert, they're good.
